src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala

```diff
@@ -129,16 +129,10 @@ trait TreeAndTypeAnalysis extends Debugging {
       def filterAndSortChildren(children: Set[Symbol]) = {
         // symbols which are both sealed and abstract need not be covered themselves,
         // because all of their children must be and they cannot otherwise be created.
-        val children1 = children.toList.filter { child =>
-          !child.isSealed ||           // keep non-sealed children
-          !child.isAbstractClass ||    // keep sealed, non-abstract children
-          isPrimitiveValueClass(child) // keep sealed, abstract, primitive classes
-        }.sortBy(_.sealedSortName)
-        children1.filter { child =>
-          !child.isAbstractClass || // keep non-abstract children
-          !children1.exists(sym => (sym ne child) && sym.isSubClass(child))
-          // keep abstract children that aren't superclasses of other children
-          // for example in t6159 drop X2
+        val children1 = children.toList.filterNot(child => child.isSealed && child.isAbstractClass).sortBy(_.sealedSortName)
+        children1.filterNot { child =>
+          // remove abstract children that are superclasses of other children, for example in t6159 drop X2
+          child.isAbstractClass && children1.exists(sym => (sym ne child) && sym.isSubClass(child))
```

Comment:
maybe we can only remove private children here? This example (two files) no longer shows an exhaustivity warning:

```
object A {
  sealed trait X
  class X1 extends X2 with X { }
  abstract class X2 extends X
  sealed trait X3 extends X

  def f(x: X) = x match {
    case _: X1 => 0
  }
}
```

```
object B {
  def main(args: Array[String]): Unit = {
    A.f(new A.X2 { })
  }
}

```